File: src/main/java/spixie/ControllerStage.java
package spixie;

import javafx.beans.value.ChangeListener;
import javafx.beans.value.ObservableValue;
import javafx.event.EventHandler;
import javafx.geometry.Orientation;
import javafx.scene.control.Button;
import javafx.scene.control.ListView;
import javafx.scene.control.SplitPane;
import javafx.scene.control.ToolBar;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.BorderPane;

public class ControllerStage extends BorderPane {
    public ControllerStage() {
        super();
        SplitPane splitPane = new SplitPane();

        final ListView<Component> componentListView = new ListView<>();
        splitPane.setResizableWithParent(componentListView, false);
        splitPane.setOrientation(Orientation.HORIZONTAL);

        final BorderPane componentBodyContainer = new BorderPane();

        componentListView.getItems().addAll(Main.world.root);
        componentListView.getSelectionModel().selectedItemProperty().addListener(new ChangeListener<Component>() {
            public void changed(ObservableValue<? extends Component> observableValue, Component component, Component t1) {
                componentBodyContainer.setCenter(componentListView.getSelectionModel().getSelectedItem().componentBody);
            }
        });


        componentListView.getSelectionModel().select(0);


        splitPane.getItems().addAll(componentListView, componentBodyContainer);
        splitPane.setDividerPosition(0,0.15);

        setCenter(splitPane);

        ToolBar menuBar = new ToolBar();
        Button renderButton = new Button("Render");
        renderButton.setOnMouseClicked(new EventHandler<MouseEvent>() {
            @Override
            public void handle(MouseEvent mouseEvent) {
                ControllerStage.this.setDisable(true);
                Main.world.renderToFile(new World.FrameRenderedToFileEvent() {
                    @Override
                    public void handle(int currentFrame, int framesCount) {
                        renderButton.setText(currentFrame + " / " + framesCount);
                    }
                }, new World.RenderToFileCompleted() {
                    @Override
                    public void handle() {
                        renderButton.setText("Render");
                        ControllerStage.this.setDisable(false);
                    }
                });
            }
        });
        menuBar.getItems().addAll(renderButton,Main.world.frame);
        setTop(menuBar);
    }
}

File: src/main/java/spixie/ControllerStage.kt
package spixie

import javafx.event.EventHandler
import javafx.geometry.Orientation
import javafx.scene.control.*
import javafx.scene.input.MouseEvent
import javafx.scene.layout.BorderPane
import spixie.components.Multiplier

class ControllerStage : BorderPane() {
    init {
        val splitPane = SplitPane()

        val componentTreeView = TreeView<ComponentsListItem>()
        SplitPane.setResizableWithParent(componentTreeView, false)
        splitPane.orientation = Orientation.HORIZONTAL

        val componentBodyContainer = BorderPane()

        componentTreeView.root = TreeItem()
        componentTreeView.isShowRoot = false
        componentTreeView.root.children.add(Main.world.root)
        componentTreeView.selectionModel.selectedItemProperty().addListener { _, _, _ -> componentBodyContainer.center = componentTreeView.selectionModel.selectedItem.value.getPropsPane() }


        componentTreeView.selectionModel.select(0)

        var contextMenu = ContextMenu()

        componentTreeView.setOnContextMenuRequested { event ->
            val selectedTreeItem = componentTreeView.selectionModel.selectedItem
            val selectedValue = selectedTreeItem.value
            when(selectedValue){
                is Component -> {
                    contextMenu.hide()
                    contextMenu = ContextMenu()
                    val menuInsert = Menu("Insert")

                    val menuItem1 = MenuItem("Multiplier")
                    menuInsert.items.addAll(menuItem1)

                    contextMenu.items.addAll(menuInsert)


                    menuItem1.setOnAction { event ->
                        val component = Multiplier()
                        selectedTreeItem.children.addAll(
                                TreeItem<ComponentsListItem>(
                                        ComponentObject(component, component.genPropsPane())
                                )
                        )
                    }
                    contextMenu.show(componentTreeView,event.screenX, event.screenY)
                }
            }
        }


        splitPane.items.addAll(componentTreeView, componentBodyContainer)
        splitPane.setDividerPosition(0, 0.15)

        center = splitPane

        val menuBar = ToolBar()
        val renderButton = Button("Render")
        renderButton.onMouseClicked = EventHandler<MouseEvent> {
            this@ControllerStage.isDisable = true
            Main.world.renderToFile(object: World.FrameRenderedToFileEvent{
                override fun handle(currentFrame: Int, framesCount: Int) {
                    renderButton.text = currentFrame.toString() + " / " + framesCount
                }
            }, object: World.RenderToFileCompleted{
                override fun handle() {
                    renderButton.text = "Render"
                    this@ControllerStage.isDisable = false
                }
            })
        }
        menuBar.items.addAll(renderButton, Main.world.frame)
        top = menuBar
    }
}
